Plutarch/Repr/SOP.hs

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+
+module Plutarch.Repr.SOP (
+  PSOPStruct (PSOPStruct, unPSOPStruct),
+  PSOPRec (PSOPRec, unPSOPRec),
+  DeriveAsSOPStruct (DeriveAsSOPStruct, unDeriveAsSOPStruct),
+  DeriveAsSOPRec (DeriveAsSOPRec, unDeriveAsSOPRec),
+) where
+
+import Control.Arrow ((&&&))
+import Data.Kind (Type)
+import GHC.Exts (Any)
+import Generics.SOP
+import Generics.SOP qualified as SOP
+import Generics.SOP.Constraint
+
+-- I think I can use Plutarch.Prelude here? maybe
+import Plutarch.Internal.Eq
+import Plutarch.Internal.Other
+import Plutarch.Internal.PLam
+import Plutarch.Internal.PlutusType
+import Plutarch.Internal.Term
+import Plutarch.Repr.Internal
+
+newtype PSOPStruct (struct :: [[S -> Type]]) (s :: S) = PSOPStruct {unPSOPStruct :: PStruct struct s}
+newtype PSOPRec (struct :: [S -> Type]) (s :: S) = PSOPRec {unPSOPRec :: PRec struct s}
+
+class SListI (PCaseTy r struct) => PSOPStructConstraint' struct r
+instance SListI (PCaseTy r struct) => PSOPStructConstraint' struct r
+
+class (SListI struct, forall r. PSOPStructConstraint' struct r) => PSOPStructConstraint struct
+instance (SListI struct, forall r. PSOPStructConstraint' struct r) => PSOPStructConstraint struct
+
+instance (SListI2 struct, PSOPStructConstraint struct) => PlutusType (PSOPStruct struct) where
+  type PInner (PSOPStruct struct) = POpaque
+  type PCovariant' (PSOPStruct struct) = All2 PCovariant'' struct
+  type PContravariant' (PSOPStruct struct) = All2 PCovariant'' struct
+  type PVariant' (PSOPStruct struct) = All2 PCovariant'' struct
+  pcon' (PSOPStruct x) = punsafeCoerce $ pconSOPStruct x
+  pmatch' x f = pmatchSOPStruct (punsafeCoerce x) (f . PSOPStruct)
+
+instance SListI struct => PlutusType (PSOPRec struct) where
+  type PInner (PSOPRec struct) = POpaque
+  type PCovariant' (PSOPRec struct) = All PCovariant'' struct
+  type PContravariant' (PSOPRec struct) = All PCovariant'' struct
+  type PVariant' (PSOPRec struct) = All PCovariant'' struct
+  pcon' (PSOPRec x) = punsafeCoerce $ pconSOPRec x
+  pmatch' x f = pmatchSOPRec (punsafeCoerce x) (f . PSOPRec)
+
+--------------------------------------------------------------------------------
+
+instance All PEq struct => PEq (PSOPRec struct) where
+  x #== y =
+    phoistAcyclic
+      ( plam $ \x' y' ->
+          pmatch x' $ \(PSOPRec (PRec x'')) ->
+            pmatch y' $ \(PSOPRec (PRec y'')) ->
+              grecEq x'' y''
+      )
+      # x
+      # y
+
+-- NOTE/TODO: Performance has not been tested for this. I feel like this would be terribly slow
```

Comment:
Though this still is slow, it's actually faster than original implementation. Because it doesn't use `#&&`